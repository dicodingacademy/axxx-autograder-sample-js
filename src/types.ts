export type Checklist = {
  key: string;
  completed: boolean;
  reason: string;
};

export type SubmissionInfo = {
  id: number;
  submitterName: string;
  submitterUsername: string;
  submitterBirthday: string;
  quizId: number;
  courseId: number;
  rejectedCount: number;
}

export type Report = {
  submissionId: number;
  message: string;
  checklistKeys: string[];
  isPassed: boolean;
  rating: number;
}