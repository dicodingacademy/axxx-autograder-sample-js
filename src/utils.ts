import { Report, SubmissionInfo } from './types';
import * as fs from 'node:fs/promises';
import { join } from 'node:path';

export async function getSubmissionInfo(path: string): Promise<SubmissionInfo>{
  const submissionInfoFilePath = join(path, 'auto-review-config.json');
  const text = await fs.readFile(submissionInfoFilePath, 'utf8');
  const json = JSON.parse(text);

  return {
    id: json.id,
    submitterUsername: json.submitter_username,
    submitterName: json.submitter_name,
    submitterBirthday: json.submitter_birthday,
    quizId: json.quiz_id,
    courseId: json.course_id,
    rejectedCount: json.rejected_count,
  };
}

export async function writeReportJson(report: Report, submissionPath: string): Promise<void> {
  const payload = {
    submission_id: report.submissionId,
    message: report.message,
    checklist_keys: report.checklistKeys,
    is_passed: report.isPassed,
    rating: report.rating,
    is_draft: false,
  };

  const reportPath = join(submissionPath, 'report.json');

  await fs.writeFile(reportPath, JSON.stringify(payload), 'utf8');
}