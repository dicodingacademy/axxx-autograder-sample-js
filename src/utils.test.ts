import { describe, expect, it } from 'vitest';
import { join } from 'node:path';
import { getSubmissionInfo, writeReportJson } from './utils';
import * as fs from 'node:fs/promises';
import { Report } from './types';

function getSubmissionFixturePath(...folders: string[]) {
  return join(process.cwd(), 'fixtures', 'submissions', ...folders);
}

function readReportJson(submissionPath: string) {
  const reportPath = join(submissionPath, 'report.json');
  return fs.readFile(reportPath, 'utf8');
}

describe('utils', () => {
  describe('getSubmissionInfo', () => {
    it('should return `SubmissionInfo` correctly', async () => {
      // Arrange
      const path = getSubmissionFixturePath('not-contain-package-json');

      // Action
      const submissionInfo = await getSubmissionInfo(path);

      // Assert
      expect(submissionInfo).toMatchSnapshot();
    });
  });

  describe('writeReportJson', () => {
    it('should write report.json correctly', async () => {
      // Arrange
      const path = getSubmissionFixturePath('not-contain-package-json');
      const report: Report = {
        submissionId: 123,
        message: '<p>Selamat! Submission Anda diterima</p>',
        isPassed: true,
        checklistKeys: ['checklist-a', 'checklist-b'],
        rating: 5,
      };

      // Action
      await writeReportJson(report, path);

      // Assert
      const reportJson = await readReportJson(path);
      expect(reportJson).toMatchSnapshot();
    });
  });
});