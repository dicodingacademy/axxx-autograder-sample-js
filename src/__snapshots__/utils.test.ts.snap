// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > getSubmissionInfo > should return \`SubmissionInfo\` correctly 1`] = `
{
  "courseId": 610,
  "id": 667273,
  "quizId": 32712,
  "rejectedCount": 2,
  "submitterBirthday": "2020-01-01 00:00:00",
  "submitterName": "Dimas Saputra",
  "submitterUsername": "dimassaputra",
}
`;

exports[`utils > writeReportJson > should write report.json correctly 1`] = `"{"submission_id":123,"message":"<p>Selamat! Submission Anda diterima</p>","checklist_keys":["checklist-a","checklist-b"],"is_passed":true,"rating":5,"is_draft":false}"`;
